### ISSUE TITLE
Add data augmentation.

### ISSUE THREAD

**Issue by brudfors** [balbasty/nitorch#40]
Data augmentation is useful in general for training NNs. I propose implementation of a new ImageAugmentor class that allows for augmenting images. But @balbasty might want it more general, so maybe start with Augmentor as the base class? This object could then be imported into a data iterator that loads batches for training/validation/testing.

My questions are:

1. Where should we put this object?
2. What type of augmentation should we include? I guess, flips, affine transformations, nonlinear transformations, intensity transformations, etc.

**Comment by balbasty** [comment]
There's preliminary stuff in [nn.generators](https://github.com/balbasty/nitorch/tree/master/nitorch/nn/generators). 

Currently, they use the DL convention, so the base class is `Module`, and they have a `forward` method that take tensors whose first dimension must be a batch dimension. The tricky thing is that parameters of the distribution we sample from can be thought of as parameters of the layer (so defined at construction and stored as attributes in the layer) or as inputs to the forward function (and therefore with a batch dim). What I also wanted is that parameters can themselves be random, and sampled during the forward pass. I just had a look at [`RandomFieldSample`](https://github.com/balbasty/nitorch/blob/893b8e8e992458808c2aabaf4e9953da380e70d0/nitorch/nn/generators/_field.py#L10) and I am not sure it is perfect (e.g., you can't have a mean that has a batch dimension).

There's still some thinking to do. You should also have a look to [torch.distributions](https://pytorch.org/docs/stable/distributions.html): it's quite well designed and you will probably have to use it.

Then there's the question of how to plug them in the trained model, we can have chat about that but we could think of:
1. the user creates a wrapping class that calls the generator first and the trained model second:
```python
class AugmentedSegNet(Module):
    def __init__(self, segnetparam, augmentationparam):
        self.augmenter = AugmentationModule(augmentationparam)
        self.segmenter = SegNet(segnetparam)

    def forward(x, ref, *, _loss=None, _metric=None):
        x, ref = self.augmenter(x, ref)
        pred = self.segmenter(x, ref, _loss=_loss, _metric=_metric)    # that's the trick I told you about to defer loss computation
       return pred
```

2. we handle augmentation within our model classes:
```python
class SegNet(Module):
    def __init__(self, *segnetparam):
        ...

    def add_augmenter(self, augmenter):
        # here augmenter is a pre-instantiated augmentation module
        self.augmenters.append(augmenter)

    def forward(x, ref, *, _loss=None, _metric=None):
        for augmenter in self.augmenters:
            x, ref = augmenter(x, ref)
        ...
        return pred
```

3. we handle it in the trainer class
```python
for batch in data_loader:
    ...
    for augmenter in self.augmenters:
        batch = augmenter(*batch)
    output = self.model(*batch, _loss=loss, _metric=metric)
    ...
```

4. we just let the users implement the augmentation in their data loader

I am sure there are other possibilities too. I kind of like the idea that the augmentation classes are modules, because you could think of models where the augmentation is really built-in and cannot be avoided (like E's stuff). I think I tend towards a mixture of 2 and 3. Note that PyTorch's models can be switched between train/eval modes and have different behaviours in both modes (that's the module developer that decides, although some built-in layers do change their behaviour depending on this state -- like the dropout layers). You could think of having something similar for the augmentation, where the behaviour in train and eval (or even test) mode is not the same (augment only in train mode).

The thing to be careful of is there are many different ways to augment (sometimes you augment both the input and reference together -- say for spatial orientation while at other times you would only augment the input but not the reference).
